types/recommendation.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2019 The MayaData Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CStorPoolClusterRecommendation is a kubernetes custom
+// resource that represents recommended configurations
+// for one given input/requirement.
+type CStorPoolClusterRecommendation struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec           CStorPoolClusterRecommendationSpec        `json:"spec"`
+	RequestDetails CStorPoolClusterRecommendationRequestSpec `json:"requestDetails"`
```

Comment:
I guess, this is no more required, since these schema will be used by inline method invocations. This is indicating design smell.

Reply:
Good to still keep this for reference. Will help to uniquely identify the recommendation if we want to store them in a cache for faster retrieval.

Reply:
@shovanmaity should the variable be named RequestSpec instead of RequestDetails ?

Reply:
The updated workflow will be like this there will be 3 calls to provision one pool -
1. Get minimum and maximum available capacity details by selecting a raid group.
2. Get the recommendation by giving capacity.
3. Users can edit one recommendation and use that to provision one CStorPoolCluster. So the recommendation need at least name, namespace, raid group configuration which will be in 2nd inline call. We can avoid copying full spec and put those fields.

Please provide suggestions. cc @AmitKumarDas  @avishnu

Reply:
I guess, lets decide based on how code shapes up. @shovanmaity you are the best person to decide. Once the relevant PR is ready, we can pitch in.

Reply:
```go
// CStorPoolClusterRecommendationSpec contains recommended
// block devices topology (blockdevices group by node)
type CStorPoolClusterRecommendationSpec struct {
	BlockDeviceTopology []BlockDeviceTopology `json:"blockDeviceTopology"`
	// Name represents name of the cspc.
	Name string `json:"name"`
	// Namespace represents namespace of the cspc.
	Namespace string `json:"namespace"`
	// DataConfig represents raid configuration for data devices.
	DataConfig RaidGroupConfig `json:"dataConfig"`
	// WriteCacheConfig represents raid configuration for write cache devices.
	// If this field is nil then write cache is disabled.
	WriteCacheConfig *RaidGroupConfig `json:"writeCacheConfig"`
	// ReadCacheConfig represents raid configuration for read cache devices.
	// If this field is nil then read cache is disabled.
	ReadCacheConfig *RaidGroupConfig `json:"readCacheConfig"`
}
```

Reply:
cc @AmitKumarDas  @avishnu  ^^^

Reply:
Let us avoid Name & Namespace in Request Spec. Let it take up the Name & Namespace of Request itself. Name can be provided by user or auto-generated. Namespace can be provided by user or it can use DAO / OpenEBS installed namespace (need to be very flexible here)

Similarly, can we rename the Name & Namespace of recommendation _(that will be same as Request)_. In future, for a list of recommendations per request, let each item have same name & namespace. Since only one recommendation will be applied at the end of the day, it should not matter.